File: day9/python/stream.py
#!/usr/bin/python3

import sys

def removeGarbage(stream):
    hasGarbage = lambda s: s.find('<') > 0

    stream = stream.replace('!!', '')
    totalGarbageLength = 0

    while hasGarbage(stream):
        begin = stream.find('<')
        end = stream.find('>')
        while stream[end - 1] == '!' and stream[end - 2] != '!':
            end = stream.find('>', end + 1)

        garbage = stream[begin + 1:end]
        marks = garbage.count('!')
        totalGarbageLength += len(garbage) - marks * 2
        stream = stream[0:begin] + stream[end + 1:]

    stream = stream.replace(',', '')
    print('Removed garbage length:', totalGarbageLength)
    return stream

def calculateScore(stream):
    score = 0
    current = 1
    for i in range(0, len(stream)):
        if stream[i] == '{':
            score += current
            current += 1
        elif stream[i] == '}':
            current -= 1
    return score

if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    stream = open(sys.argv[1]).read()
    stream = removeGarbage(stream)
    print('Score:', calculateScore(stream))

File: day09/rust/stream.rs
use std::fs;

fn remove_garbage(mut stream: String) -> String {
    let has_garbage = |s: &str| {
        if let Some(_) = s.find("<") {
            return true;
        }
        false
    };

    stream = stream.replace("!!", "");

    let mut garbage_length = 0;
    while has_garbage(&stream) {
        let begin = stream.find("<").unwrap();
        let mut end = stream.find(">").unwrap();
        while stream.chars().nth(end - 1).unwrap() == '!' {
            end += (&stream[end + 1..]).find(">").unwrap() + 1;
        }
        let marks = stream[begin..=end].matches('!').count();
        stream.replace_range(begin..=end, "");
        garbage_length += end - begin - 1 - 2 * marks;
    }
    println!("{}", garbage_length);
    stream
}

fn calculate_score(stream: &str) -> u32 {
    let mut total_score = 0;
    let mut current_value = 0;

    for c in stream.chars() {
        if c == '{' {
            current_value += 1;
            total_score += current_value;
        } else if c == '}' {
            current_value -= 1;
        }
    }
    total_score
}

fn main() {
    let stream = fs::read_to_string("../input").expect("file not found");
    println!("{}", calculate_score(&remove_garbage(stream)));
}
